python/ray/data/_internal/table_block.py

```diff
@@ -46,6 +48,20 @@ def add(self, item: Union[dict, TableRow, np.ndarray]) -> None:
                 "Returned elements of an TableBlock must be of type `dict`, "
                 "got {} (type {}).".format(item, type(item))
             )
+
```

Comment:
Should add_block() be schema checked as well?

Reply:
It seems we already have actionable error message:

```
def bad_fn(row):
  if random.random() > 0.5:
    return pa.Table.from_pydict({"a": ["hello1"]})
  else:
    return pa.Table.from_pydict({"a": [1]})
    
ds = ray.data.range(10).repartition(1).map_batches(bad_fn, batch_size=1)
```

```
>>> 2022-10-05 17:25:13,296	ERROR worker.py:400 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::_map_block_nosplit() (pid=80558, ip=127.0.0.1)
  File "/Users/chengsu/ray/python/ray/data/_internal/compute.py", line 439, in _map_block_nosplit
    for new_block in block_fn(block, *fn_args, **fn_kwargs):
  File "/Users/chengsu/ray/python/ray/data/dataset.py", line 546, in transform
    yield output_buffer.next()
  File "/Users/chengsu/ray/python/ray/data/_internal/output_buffer.py", line 74, in next
    block = self._buffer.build()
  File "/Users/chengsu/ray/python/ray/data/_internal/delegating_block_builder.py", line 64, in build
    return self._builder.build()
  File "/Users/chengsu/ray/python/ray/data/_internal/table_block.py", line 101, in build
    return self._concat_tables(tables)
  File "/Users/chengsu/ray/python/ray/data/_internal/arrow_block.py", line 115, in _concat_tables
    return pyarrow.concat_tables(tables, promote=True)
  File "pyarrow/table.pxi", line 2338, in pyarrow.lib.concat_tables
  File "pyarrow/error.pxi", line 143, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Unable to merge: Field a has incompatible types: int64 vs string
```

`add_block()` is different from `add()` as it just stores Arrow blocks in a list, and I think we should be good now.